managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2954,53 +2971,99 @@ private static List<StringProperty> buildStringPropertiesMap(Map<String, String>
         return stringProperties;
     }
 
-    private List<MLDataFormats.MessageRange> buildIndividualDeletedMessageRanges() {
+    @VisibleForTesting
+    List<MLDataFormats.MessageRange> buildIndividualDeletedMessageRanges() {
```

Comment:
Instead of adding "VisibleForTesting" annotation here, please extract this logic to a separate class which is tested separately. There's a lot of complexity here and this would be very hard to maintain in the future in this form. Smaller classes and smaller methods are a general preference in maintainable code.